lora/sx127x/sx127x.go

```diff
@@ -0,0 +1,344 @@
+// Package lora provides a driver for SX127x LoRa transceivers.
+//
+// Datasheet:
+// https://www.semtech.com/uploads/documents/DS_SX1276-7-8-9_W_APP_V6.pdf
+//
+// LoRa Configuration Parameters:
+//
+// Frequency: is the frequency the tranceiver uses. Valid frequencies depend on
+//  the type of LoRa module, typically around 433MHz or 866MHz. It has
+//  a granularity of about 23Hz, how close it can be to others depends on the
+//  Bandwidth bing used.
+//
+// Bandwidth: is the bandwidth used for tranmissions, ranging from 7k8 to 512k
+//  A higher bandwidth gives faster transmissions, lower gives greater range
+//
+// SpreadingFactor: is how a transmission is spread over the spectrum. It ranges
+//  from 6 to 12, a higher value gives greater range but slower transmissions.
+//
+// CodingRate: is the cyclic error coding used to improve the robustness of the
+//  transmission. It ranges from 5 to 8, a higher value gives greater
+//  reliability but slower transmissions.
+//
+// TxPower: is the power used for the transmission, ranging from 1 to 20.
+//  A higher power gives greater range. Regulations in your country likely
+//  limit the maximum power permited.
+//
+// Presently this is only synchronous and so does not use any DIOx pins
+//
+package sx127x
+
+import (
+	"errors"
+	"fmt"
+	"machine"
+	"time"
+)
+
+// Device wraps an SPI connection to a SX127x device.
+type Device struct {
+	spi    machine.SPI
+	csPin  machine.GPIO
+	rstPin machine.GPIO
+}
+
+type Config struct {
+	Frequency       uint32
+	SpreadingFactor uint8
+	Bandwidth       int32
+	CodingRate      uint8
+	TxPower         int8
+}
+
+// New creates a new SX127x connection. The SPI bus must already be configured.
+func New(spi machine.SPI, csPin machine.GPIO, rstPin machine.GPIO) Device {
+	return Device{
+		spi:    spi,
+		csPin:  csPin,
+		rstPin: rstPin,
+	}
+}
+
+// Configure initializes and configures the LoRa module
+func (d *Device) Configure(cfg Config) (err error) {
+	d.csPin.High()
+
+	d.reset()
+
+	if d.readRegister(REG_VERSION) != 0x12 {
+		return errors.New("SX127x module not found")
+	}
+
+	d.sleep()
+
+	// set base addresses
+	d.writeRegister(REG_FIFO_TX_BASE_ADDR, 0)
+	d.writeRegister(REG_FIFO_RX_BASE_ADDR, 0)
+
+	// set LNA boost
+	d.writeRegister(REG_LNA, d.readRegister(REG_LNA)|0x03)
+
+	// set auto AGC
+	d.writeRegister(REG_MODEM_CONFIG_3, 0x04)
+
+	err = d.ReConfigure(cfg)
+
+	d.idle()
+
+	return err
+}
+
+// ReConfigure updates the LoRa module configuration
+func (d *Device) ReConfigure(cfg Config) (err error) {
```

Comment:
I think this should be `Reconfigure()`.

Reply:
None of the other drivers have anything like this. Why not expose the specific setters, like `SetFrequency` etc.? It makes it more obvious how to set these properties. It's not immediately obvious from the name which things are configured with `ReConfigure` and which are not.

Reply:
Maybe not important...
Not all parameters can be changed in all modes, for example setFreq would work only in Idle, if I remember correctly.